### PR TITLE
Do not generate new random number while receiving HRR

### PR DESCRIPTION
**Pull request by BensonLiou**
## Description

According to [RFC8446](https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.2), random number of 2nd clientHello should be same as first one while receiving HRR. This PR is to preventing re-generating random number on 2nd clientHello
Fix [8669](https://github.com/Mbed-TLS/mbedtls/issues/8669)

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided, or not required
- [x] **backport** done, or not required
- [x] **tests** provided, or not required

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Thanks for contributing a fix! Please amend your commit to [add a sign-off line](https://stackoverflow.com/questions/13457203/how-to-add-the-signed-off-by-field-in-the-git-patch). We need the DCO check to pass to accept external contributions.

We normally want a non-regression test for bug fixes. Would you mind adding one? I guess it would be a unit test in `test/suites/test_suite_ssl` that does two 1.3 handshakes, arranges for the server to send a HelloRetryRequest, records the random values sent by the client and checks that they're the same.

**Comment by BensonLiou** [comment]
I will do it but I might need time to understand how to add test cases to the framework. It will be helpful if there are documents to describe it.

**Comment by gilles-peskine-arm** [comment]
There's general guidance in https://mbed-tls.readthedocs.io/en/latest/kb/development/test_suites/ though you can probably figure most of it by example looking at `tests/suites/test_suite_ssl.data` and `tests/suites/test_suite_ssl.function`.

For how to write the test cases, I'm not very familiar with the helper functions we have for SSL, but I think a good start would be the `move_handshake_to_state` function, with some added code to have the server send ServerRetryRequest, and compare the random read from the handshake structure before and after the client processes the SRR.

**Comment by gilles-peskine-arm** [comment]
Please avoid including merge commits in pull requests, unless a conflict appears after the first review. It's ok if a pull request forks from an old commit as long as there's no conflict.

**Comment by BensonLiou** [comment]
The 2nd merge is required because I need the test function **tls13_cli_early_data_status()**
I check the random number from client side instead of server side since client state is more controllable and also less code change.

**Comment by gilles-peskine-arm** [comment]
Thanks for the update! We are two days from the 3.6.0 release code freeze, we'll try to include this fix.

One thing I can see before reviewing in detail is that this is missing a [changelog entry](https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md). Something like:

```
Bugfix
   * In TLS 1.3 clients, fix an interoperability problem due to the client
     generating a new random after a HelloRetryRequest. Fixes #8669.
```

For future reference, when a pull request hasn't been reviewed yet, we prefer to rebase on top of `development` rather than merge (but only if there are conflicts, if not a pull request forked from an older commit is fine), because merges take work to review. But at this point here, please keep the existing history with the merges.

**Comment by gilles-peskine-arm** [comment]
Unfortunately there's a merge conflict again :cry:

**Comment by BensonLiou** [comment]
I use the github CLI to resolve the conflict and it seems github do the merge for me. Sorry for that😣

**Comment by gilles-peskine-arm** [comment]
Thanks! We still need a changelog entry though.

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks for the updates! Looks good to me now.

**Review by valeriosetti — DISMISSED** [review]
LGTM if the CI is OK

File: library/ssl_client.c [hunk 1]
```diff
@@ -792,10 +792,15 @@ static int ssl_prepare_client_hello(mbedtls_ssl_context *ssl)
         (ssl->handshake->cookie == NULL))
 #endif
     {
-        ret = ssl_generate_random(ssl);
-        if (ret != 0) {
-            MBEDTLS_SSL_DEBUG_RET(1, "Random bytes generation failed", ret);
-            return ret;
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
+        if (ssl->handshake->hello_retry_request_count == 0)
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/ssl_client.c]: In the interest of time, I've taken the liberty of pushing this fix, so that we can get a CI run.

**Review by valeriosetti — APPROVED** [review]

**Review by tom-daubney-arm — APPROVED** [review]
I have reviewed only commit 93b305d as requested by @gilles-peskine-arm.

So I think we have two full reviews in the aggregate now. @gilles-peskine-arm up to but not including 93b305d and me on 93b305d and then @valeriosetti has done a full review.

**Comment by gilles-peskine-arm** [comment]
Thanks everyone for the quick reactions! We'll have this fixed in the 3.6.0 release.

**Pull request merged**